### ISSUE TITLE
File directive can access out-of-sandbox files

### ISSUE THREAD

**Issue by chmeliik** [pypa/setuptools#2296]
Found while examining setuptools code, I can submit a PR to fix both issues described below if desired.

## Scenario 1 - `../$PWD*/`

Directory structure:

```
.
├── foo
│   ├── setup.cfg
│   └── setup.py
└── foobar
    └── file.txt
```

setup.cfg:

```ini
[metadata]
name = foo
version = file:../foobar/file.txt
```

Result:

```bash
$ # setup.py is just 'import setuptools; setuptools.setup()'
$ python setup.py sdist

Traceback (most recent call last):
...
distutils.errors.DistutilsOptionError: Version loaded from file: ../foobar/file.txt does not comply with PEP 440: I'm a file
```

Expected is that the file would have been rejected for being outside the project root, but instead, the file was read.

Because of the way setuptools does the `_assert_local()` check, the `file:` directive under `foo/` can access any file under `../foo*/`

## Scenario 2 - symlink

Directory structure:

```
foo
├── setup.cfg
├── setup.py
└── passwd -> /etc/passwd
```

setup.cfg:

```ini
[metadata]
name = foo
version = file: passwd
```

Result:

```bash
$ # setup.py is just 'import setuptools; setuptools.setup()'
$ python setup.py sdist

Traceback (most recent call last):
...
distutils.errors.DistutilsOptionError: Version loaded from file: passwd does not comply with PEP 440: <omitted>
```

Expected is that the file would have been rejected as its symlink target ultimately resolves outside of the project root.

`os.path.abspath()` does not resolve symlinks, `os.path.realpath()` should be used instead

**Comment by jaraco** [comment]
I've update your original report to include the `file:` in the first scenario and to describe the missed expectation in each scenario.

**Comment by jaraco** [comment]
For Scenario 1, while I agree the behavior is surprising, it seems unlikely to me to be a serious concern. Is this a theoretical issue or one you encountered in the wild? Regardless, I'm open to being more precise about protecting the user here, so do feel free to submit a PR for consideration.

Regarding Scenario 2, I'm less convinced the issue you present is an issue. I suspect there are legitimate cases of using a symlink to reference content outside of the project root. It seems reasonable that if the project/user wishes to use a symlink to make external content appear in the project root, that still honors the expectation.

In your example, `/etc/passwd` is accessed, but the project author could just as easily have invoked behavior to `cp /etc/passwd passwd` before invoking the build.

I'm not convinced this behavior is illegitimate, but I'm open to being convinced. Changing this behavior would probably be considered backward-incompatible and might require a deprecation period to migrate existing projects off of relying on this behavior (if they are).

Would you be willing to shepherd the project and its users through this transition? Can you make a case that there shouldn't be a transition (that a hard switch should be safe for all users)? If yes to either, please feel free to draft a PR. This change in particular probably demands a test to capture the failed expectation.

**Comment by chmeliik** [comment]
I agree that the 1st scenario is a very niche corner case, it probably cannot ever cause real issues.

I thought the 2nd scenario would be the more interesting one here, since it can theoretically be used to access any file on the build machine, but in the typical case where "build machine" = the user's PC, this is a non-issue (and if you are concerned about the build machine's safety, then arbitrary code execution from setup.py or the `attr:` directive is obviously a much more serious issue).

I needed to replicate some of setuptools' functionality while avoiding arbitrary code execution, which is how I came across this potential issue, if you see it as a legitimate use case that is completely fine by me. Closing issue.
